File: hackerrank/easy/0049-halloween-sale/swift/solution.swift
// solution.swift

import Foundation

func howManyGames(p: Int, d: Int, m: Int, s: Int) -> Int {
    var gameCount = 0
    var budget = s
    var currentPrice = p

    while budget >= currentPrice {
        budget -= currentPrice
        gameCount += 1
        currentPrice = max(currentPrice - d, m)
    }

    return gameCount
}

// --- demo run ---
let sample1 = howManyGames(p: 20, d: 3, m: 6, s: 80)
print(sample1) // Expected: 6

let sample2 = howManyGames(p: 20, d: 3, m: 6, s: 85)
print(sample2) // Expected: 7

let sample3 = howManyGames(p: 20, d: 3, m: 6, s: 23)
print(sample3) // Small budget test

File: hackerrank/easy/0049-halloween-sale/python/solution.py
## solution.py

from typing import Final


def howManyGames(p: int, d: int, m: int, s: int) -> int:
    game_count = 0
    budget = s
    current_price = p

    while budget >= current_price:
        budget -= current_price
        game_count += 1
        current_price = max(current_price - d, m)

    return game_count


# --- demo run ---
if __name__ == "__main__":
    print(howManyGames(20, 3, 6, 80))  # Expected: 6
    print(howManyGames(20, 3, 6, 85))  # Expected: 7
    print(howManyGames(20, 3, 6, 23))  # Small budget test
